File: 1436-get-watched-videos-by-your-friends/1436-get-watched-videos-by-your-friends.cpp
class Solution {
public:
    static bool comparator(pair<string,int>& a,pair<string,int>& b) {
        if( a.second==b.second) {
            return a.first<=b.first;
        }
        return a.second<=b.second;
    }
    vector<string> watchedVideosByFriends(vector<vector<string>>& watchedVideos, vector<vector<int>>& friends, int id, int level) {
        int n = friends.size();
        unordered_map<int,int> visited;
        queue<pair<int,int>> q;
    
        q.push({id,0});
        vector<int> levelKFriends;
        visited[id]=1;
        while(!q.empty()) {
            int personId = q.front().first;
            int personLevel = q.front().second;
            q.pop();
            for(auto person:friends[personId]) {
                if(visited.find(person)==visited.end()) {
                    if(personLevel+1==level) {
                        levelKFriends.push_back(person);
                        visited[person]=1;
                    } else {
                        q.push({person,personLevel+1});
                        visited[person]=1;
                    }
                }
            }
        }

        unordered_map<string,int> freq;
        for(auto person:levelKFriends) {
            for(auto watchedVideo:watchedVideos[person]) {
                freq[watchedVideo]++; 
       }
        }
        vector<pair<string,int>> watchedVideoVector;
        for(auto video:freq) {
            watchedVideoVector.push_back({video.first,video.second});
        }
        sort(watchedVideoVector.begin(),watchedVideoVector.end(),comparator);
        vector<string> ans;
        for(auto watchedVideo:watchedVideoVector) ans.push_back(watchedVideo.first);
        return ans;
    }
};

File: 1436-get-watched-videos-by-your-friends/1436-get-watched-videos-by-your-friends.py
from collections import defaultdict, deque
from typing import List

class Solution:
    def watchedVideosByFriends(self, watchedVideos: List[List[str]], friends: List[List[int]], id: int, level: int) -> List[str]:
        watchedVideosDict = defaultdict(int)
        visited = set([id])  # Initialize with the starting person to avoid revisiting

        # BFS initialization
        queue = deque([(id, 0)])

        while queue:
            friendID, friendLevel = queue.popleft()

            # If we reach the desired level, count watched videos of this friend
            if friendLevel == level:
                for watchedVideo in watchedVideos[friendID]:
                    watchedVideosDict[watchedVideo] += 1
            # Continue the BFS only if we're below the target level
            elif friendLevel < level:
                for neighbourID in friends[friendID]:
                    if neighbourID not in visited:
                        visited.add(neighbourID)  # Mark as visited
                        queue.append((neighbourID, friendLevel + 1))

        # Sort videos by frequency first, then lexicographically by video name
        sortedVideos = sorted(watchedVideosDict.items(), key=lambda x: (x[1], x[0]))

        # Return only the video names in the sorted order
        return [video for video, _ in sortedVideos]


